### PR TITLE
Bug 1283030 - Remember the last normal/private tab when switching between the modes

### PR DESCRIPTION
**Pull request by tylerlacroix**

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -3339,12 +3339,15 @@ extension BrowserViewController: TopTabsDelegate {
...
+    func topTabsDidPressPrivateTab(tab: Tab?) {
         guard let selectedTab = tabManager.selectedTab else {
             return
         }
         urlBar.leaveOverlayMode()
-        if selectedTab.isPrivate {
+        if let tab = tab {
+            tabManager.selectTab(tab)
+        }
+        else if selectedTab.isPrivate {
```
Comment (sleroux) [inline Client/Frontend/Browser/BrowserViewController.swift]: Nit: I find this `if` statement difficult to parse. It seems weird that each of these would be an exclusive case - like selecting the tab, but then checking the selected tab if it's private.

File: Client/Frontend/Browser/TopTabsViewController.swift [hunk 2]
```diff
@@ -71,6 +71,9 @@ class TopTabsViewController: UIViewController {
         return delegate
     }()
     
+    private weak var lastNormalTab: Tab?
+    private weak var lastPrivateTab: Tab?
```
Comment (sleroux) [inline Client/Frontend/Browser/TopTabsViewController.swift]: Now that we are caching the last selected tabs for normal/private mode in the VC, we'll want to make sure to respond and act on all possible tab manipulation events. For example, when selecting 'Close all tabs' I don't thin the `didSelectedTabChange` delegate gets fired but the `didRemoveTab` one does so we'll want to make sure to implement that as well.

File: Client/Frontend/Browser/TopTabsViewController.swift [hunk 3]
```diff
@@ -173,7 +176,7 @@ class TopTabsViewController: UIViewController {
     }
     
     func privateTabTapped() {
-        delegate?.topTabsDidPressPrivateTab()
+        delegate?.topTabsDidPressPrivateTab(isPrivate ? lastNormalTab : lastPrivateTab)
```
Comment (sleroux) [inline Client/Frontend/Browser/TopTabsViewController.swift]: This... seems weird that we are passing in a normal tab as the parameter to `topTabsDidPressPrivateTab`. Either the method needs to be renamed or what you pass to it needs to changed.

Comment (tylerlacroix) [inline Client/Frontend/Browser/TopTabsViewController.swift]: Yeah, true, I think I need to rename this to make more sense.  Would topTabsDidTogglePrivateMode make more sense? :)

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 1 above)
Comment (tylerlacroix) [inline Client/Frontend/Browser/BrowserViewController.swift]: I think "tab" was a poor choice for a parameter name, I think "cachedTab" would make more sense?  So I check if there's a cached tab passed is the function, if not then I select the most recent normal/private tab to select.

File: Client/Frontend/Browser/TopTabsViewController.swift (hunk 2 above)
Comment (tylerlacroix) [inline Client/Frontend/Browser/TopTabsViewController.swift]: I'll look into those cases! :)

File: Client/Frontend/Browser/TopTabsViewController.swift (hunk 3 above)
Comment (sleroux) [inline Client/Frontend/Browser/TopTabsViewController.swift]: I think `didTogglePrivateMode` works.

**Comment by tylerlacroix** [comment]
This should also be good for another look! :)

**Comment by sleroux** [comment]
Looks good - just need a rebase before I can merge it in.

**Pull request merged**